icing/icing_optimisationProofsScript.sml

```diff
@@ -603,11 +603,376 @@ Proof
   \\ fs[]
 QED
 
+fun rename_each [] = ALL_TAC
+| rename_each [first] = qmatch_asmsub_rename_tac first \\ qpat_x_assum first mp_tac \\ disch_then assume_tac
+| rename_each (first::rest) = qmatch_asmsub_rename_tac first \\ qpat_x_assum first mp_tac \\ rename_each rest \\ disch_then assume_tac
+
+
 Theorem fp_assoc_gen_correct:
   ∀ fpBop st1 st2 env e r.
     is_rewriteFPexp_correct [fp_assoc_gen fpBop] st1 st2 env e r
 Proof
-  cheat
+  rw[is_rewriteFPexp_correct_def]
+  \\ REVERSE (qspecl_then [‘e’, ‘fpBop’] strip_assume_tac fp_assoc_gen_cases)
+  >- (
+  fs[]
+  \\ extend_eval_tac ‘evaluate st1 _ _ = _’ ‘[fp_assoc_gen fpBop]’
+  \\ strip_tac
+  \\ pop_assum (mp_then Any mp_tac (CONJUNCT1 evaluate_add_choices))
+  \\ disch_then (qspec_then ‘st1.fp_state.choices’ assume_tac)
+  \\ fsrw_tac [SATISFY_ss] []
+  )
+  \\ rveq
+  \\ fs[]
+  \\ qpat_x_assum ‘_ = (_, _)’ (mp_tac o SIMP_RULE std_ss [evaluate_def])
+  \\ simp[REVERSE_DEF, astTheory.getOpClass_def, astTheory.isFpBool_def,
+          Once evaluate_cons, evaluate_case_case]
+  \\ disch_then (mp_tac o (SIMP_RULE std_ss [evaluate_def]))
+  \\ simp[REVERSE_DEF, astTheory.getOpClass_def, astTheory.isFpBool_def,
+          Once evaluate_cons, evaluate_case_case]
+  \\ ntac 2 (TOP_CASE_TAC \\ fs[])
+  \\ imp_res_tac evaluate_sing
+  \\ rveq
+  \\ fs[]
+  \\ rveq
+  \\ rename1 ‘evaluate st1 env [e3] = (st1N, Rval [v3])’
+  \\ ‘st1N.fp_state.canOpt = FPScope Opt’ by fp_inv_tac
+  \\ simp[Once evaluate_cons, evaluate_case_case]
+  \\ disch_then (mp_tac o (SIMP_RULE std_ss [evaluate_def]))
+  \\ simp[REVERSE_DEF, astTheory.getOpClass_def, astTheory.isFpBool_def,
+          Once evaluate_cons, evaluate_case_case]
+  \\ ntac 2 (TOP_CASE_TAC \\ fs[])
+  \\ imp_res_tac evaluate_sing
+  \\ rveq
+  \\ fs[]
+  \\ rveq
+  \\ rename1 ‘evaluate st1N env [e2] = (st2N, Rval [v2])’
+
+  (* Previously, v3 is renamed to v' by one of the rveq. We rename it back. *)
+  \\ qpat_x_assum ‘evaluate _ _ [e2] = (_, _)’ mp_tac
+  \\ rename1 ‘evaluate st1 env [e3] = (st1N, Rval [v3])’
+  \\ disch_then assume_tac
+
+  \\ simp[do_app_def, CaseEq"option", CaseEq"v", CaseEq"prod", CaseEq"result"]
+  \\ rpt strip_tac
+  \\ rveq
+  \\ fs[]
+  \\ rveq
+  \\ imp_res_tac evaluate_sing
```

Comment:
This line is a good example for what I meant above with "conceptual steps". You have an assumption `evaluate _ _ [e] = (_, v)` and want to prove that `v` must be a singleton list. Thus `imp_res_tac evaluate_sing \\ rveq \\ fs[]` can be put onto a single line (in my opinion)